plugins/janus_audiobridge.c

```diff
@@ -1503,17 +1535,35 @@ void janus_audiobridge_incoming_rtp(janus_plugin_session *handle, int video, cha
 		}
 		/* Enqueue the decoded frame */
 		janus_mutex_lock(&participant->qmutex);
-		g_queue_push_tail(participant->inbuf, pkt);
-		/* Keep the queue at max of 150 packets, though */
-		if(g_queue_get_length(participant->inbuf) > 150) {
-			/* Remove the oldest packet */
-			janus_audiobridge_rtp_relay_packet *pkt = g_queue_pop_head(participant->inbuf);
-			if(pkt != NULL) {
-				if(pkt->data)
-					g_free(pkt->data);
-				pkt->data = NULL;
-				g_free(pkt);
-				pkt = NULL;
+		/* Insert packets sorting by sequence number */
+		participant->inbuf = g_list_insert_sorted(participant->inbuf, pkt, &janus_audiobridge_rtp_sort);
+		if(participant->prebuffering) {
+			/* Still pre-buffering: do we have enough packets now? */
+			if(g_list_length(participant->inbuf) == DEFAULT_PREBUFFERING) {
+				participant->prebuffering = FALSE;
+				JANUS_LOG(LOG_VERB, "Prebuffering done! Finally adding the user to the mix\n");
+			} else {
+				JANUS_LOG(LOG_VERB, "Still prebuffering (got %d packets), not adding the user to the mix yet\n", g_list_length(participant->inbuf));
+			}
+		} else {
+			/* Make sure we're not queueing too many packets: if so, get rid of the older ones */
+			if(g_list_length(participant->inbuf) >= DEFAULT_PREBUFFERING*2) {
```

Comment:
Yeah dropping the quality for specific users was what I was talking about as a possibility to investigate, e.g., as a conseguence of slow_lnk events. We don't use the audiobridge on Meetecho usually, we have a SIP server (Asterisk) for audio conferencing and so call that via SIP plugin: that conference bridge still uses Opus and is configured at 16kHz, so the bandwidth is pretty much the same. Since we did the implementation of Opus support in Asterisk and we basically reused the same encoding/decoding code in the AudioBridge plugin as well, it's safe to assume the bandwidth consumption is pretty much the same.

Audio packets don't usually consume much bandwidth, but there are 50 of them per second (20ms each). Lost packets are typically NACKed by the peer who didn't receive them, so you may want to look into the admin API info for the affected handles, as we list the number of incoming (user lost packet Janus sent) and outgoing (Janus lost packet user sent) NACKs on there as well. If you see a lot of NACKs, something's broken on the media path between the server and the user (e.g., flaky WiFi or poor connectivity in general), and dropping the quality may or may not help.